### ISSUE TITLE
motorway_link is preferred instead of motorway

### ISSUE THREAD

**Issue by scaidermern** [osmandapp/Osmand#8376]
South-west of Neuss, North Rhine-Westphalia, Germany from 51.1854744,6.6639023 to 51.1783005,6.6798592. OsmAnd leaves the motorway and enters motorway_link just to return shortly after to the same motorway.

![photo_2020-02-02_19-29-51](https://user-images.githubusercontent.com/1040784/73613089-6a893300-45f2-11ea-9ea6-a7585d262ea9.jpg)

Route: [route.gpx.txt](https://github.com/osmandapp/Osmand/files/4144899/route.gpx.txt)

Doesn't happen with [GraphHopper](https://www.openstreetmap.org/directions?engine=graphhopper_car&route=51.18624%2C6.66318%3B51.17816%2C6.68089#map=16/51.1822/6.6720&layers=N) or [OSRM](https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=51.1862%2C6.6632%3B51.1782%2C6.6809).

Note: The motorway_link doesn't have a maxspeed set. This is correct, though. There are no maxspeed signs, you can check on mapillary.

Similar to #6037.

**Comment by scaidermern** [comment]
The same happens at the north-west of Köln, Germany, from 50.9738121,6.9170227 to 50.9668125,6.9210649.

![photo_2020-02-02_19-56-55](https://user-images.githubusercontent.com/1040784/73613482-63fcba80-45f6-11ea-9236-699c0db67726.jpg)

Route: 
[route.gpx.txt](https://github.com/osmandapp/Osmand/files/4144988/route.gpx.txt)

Doesn't happen with [GraphHopper](https://www.openstreetmap.org/directions?engine=graphhopper_car&route=50.9738%2C6.9170%3B50.9650%2C6.9227) or [OSRM](https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=50.9738%2C6.9170%3B50.9650%2C6.9227).

Again, there is no maxspeed for the motorway_link. This corresponds to reality.

**Comment by scaidermern** [comment]
The same happens at the north-east of Gera, Germany, from 50.9062386,12.1191683 to 50.8984317,12.1350646.

![photo_2020-02-02_20-11-12](https://user-images.githubusercontent.com/1040784/73613662-4597be80-45f8-11ea-80d1-e1eb4c2c7fd8.jpg)

Route: [route.gpx.txt](https://github.com/osmandapp/Osmand/files/4145044/route.gpx.txt)

**Does** happen with [GraphHopper](https://www.openstreetmap.org/directions?engine=graphhopper_car&route=50.9063%2C12.1190%3B50.8980%2C12.1357) but not with [OSRM](https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=50.9063%2C12.1190%3B50.8980%2C12.1357).

Again, there is no maxspeed for the motorway_link. This corresponds to reality.

**Comment by naoliv** [comment]
When there are no signs in Germany, it's assumed some default max speed or there is really no max speed (even for links)?

**Comment by scaidermern** [comment]
> When there are no signs in Germany, it's assumed some default max speed or there is really no max speed (even for links)?

In fact, I had to perform some investigations about this. According to the German law the speed limit remains in force until there is another sign which either cancels it or indicates a new speed limit.

This means that the tagging is actually wrong. If there is no explicit speed limit sign for motorway_link ways then the speed limit of the previous motorway still applies. This needs to be fixed in the data.

However consider the following hypothetical case
![foo](https://user-images.githubusercontent.com/1040784/73651793-81c53080-4685-11ea-8e77-d6430bbcc244.png)

There is a speed limit right before the motorway junction, applying both to the motorway and the motorway_link. Shortly after the speed limit of the motorway is reduced, e.g. because of road construction. Now the speed limit of the link is significantly higher and OsmAnd will route along the link instead of staying on the motorway.

This is a special case that occurs only rarely, though.

**Comment by Discostu36** [comment]
As long as this is only a hypothetical bug, I don't think any action is needed. Does the wrong routing disappear after adding the speed limits?

**Comment by scaidermern** [comment]
> As long as this is only a hypothetical bug, I don't think any action is needed. Does the wrong routing disappear after adding the speed limits?

For the third example (near Gera) the bug persists after adding the speed limit to the link.

The above hypothetical example isn't hypothetical after all since this situation (almost) matches reality. The [link](https://www.openstreetmap.org/way/24591529#map=19/50.90535/12.12096&layers=N) has a speed limit of 120 whereas for the motorway the speed limit gets reduced to [100](https://www.openstreetmap.org/way/) and then to [80](https://www.openstreetmap.org/way/25607787) due to road works. OsmAnd continues to route along the link.

I can't compare these recently added speed limits with other routing enginces since I don't know their update interval.

**Comment by Discostu36** [comment]
OK, I guess this can lead to two conclusions:

1) The penalty for motorway → motorway_link has to be increased
2) Is it actually faster to drive along the link? If yes, this is a feature, not a bug

**Comment by scaidermern** [comment]
>     2. Is it actually faster to drive along the link? If yes, this is a feature, not a bug

I suppose it is, in this particular case. However I wouldn't consider this a feature. The link can't cope with the whole motorway traffic, so telling people to use the link because it is faster is just wrong. Also, changing from the motorway to the link and back will increase the chance of an accident and thus should be avoided if possible.

**Comment by Discostu36** [comment]
I guess you're right, but I guess this is for the devs to decide.

**Comment by IgorEliezer** [comment]
At least in my country, all motorway links that I know of are 40-80 km/h max speed ([see example on Mapillary](https://www.mapillary.com/app/?lat=-23.75585845306048&lng=-46.54208404192754&z=15&pKey=wO4A74HEAgu1Xi3MpIEkg6&focus=map&signs=true)).

But the issue is not limited to motorway links, there's this trunk link too:

![image](https://user-images.githubusercontent.com/4673289/73850721-fd9cb580-480a-11ea-8cb9-a0557e55be7d.png)
([route](https://www.openstreetmap.org/directions?engine=graphhopper_car&route=-23.66404%2C-46.50641%3B-23.66413%2C-46.51287#map=18/-23.66418/-46.50955))

The issue above is that, the trunk way is 60 km/h while the link way has no maxspeed sign. OsmAnd needs to assume a maxspeed for link ways but it should be lower than the maxspeed of main way counterparts.
- motorway/trunk = 60 km/h
- primary = 50 km/h
- secondary= 40 km/h
- tertiary = 30 km/h

**Comment by vshcherb** [comment]
Routing doesn't consider what's faster or slower to ride it calculates what's allowed speed to ride on the link and chooses route with minimal time. 
Link has a penalty of 5 extra seconds to drive in/out, so if link and road has same distance and speed limit - road will be chosen. In case link is 130/80 = 1.6 faster than road then link will be chosen in most of the cases.

Again routing engine doesn't select between 2 types of roads, it just finds the most optimal way with min time

**Comment by naoliv** [comment]
@vshcherb it seems that after https://github.com/osmandapp/OsmAnd-resources/pull/588 these types of problems became more prominent.

Maybe some penalties could be tweaked a little more?

**Comment by scaidermern** [comment]
@vshcherb Can you explain why this issue has been closed? It doesn't matter if the current routing engine doesn't distinguish between a link and a regular road. The resulting route is wrong since it leads to a dangerous behavior in my opinion. If it can't be fixed at the moment then better leave it open to be fixed in the future.

**Comment by vshcherb** [comment]
So than this bug will be reopened https://github.com/osmandapp/Osmand/issues/7306.

Basically it becomes pure data issue which needs to have extra information when link should be used as a shortcut and when it should not be. These 2 issues don't produce any result but we don't have any data solution and it is clear that nothing could be done on software side

**Comment by scaidermern** [comment]
> So than this bug will be reopened #7306.

You are right, this is a conflict. So I guess there is no simple way to fix both cases :(

**Comment by vshcherb** [comment]
In pedestrian "OSM-routing" data we have tag "use_sideway" which suggest to not use this road and give extra penalty, here we probably need "routing_speed" or "routing_priority" which will indicate that roads are same with priority  or having restriction.

**Comment by dead10ck** [comment]
@vshcherb it's not clear to me that "nothing can be done on the software side." Other routers don't have these issues. I can appreciate that this is a difficult problem, but I think it might be more productive to acknowledge that this problem exists. It is more than fair and reasonable to say you don't have the time or resources to figure out a solution that works for both of these cases, but I don't think it's reasonable to just say "nothing can be done."

**Comment by vshcherb** [comment]
The proper solution is on the map data side and that's how it should be approached in that case top one

**Comment by vshcherb** [comment]
https://github.com/osmandapp/Osmand/issues/9436

**Comment by vshcherb** [comment]
This exactly what happened with road of 80kmh
https://www.openstreetmap.org/way/25607787#map=17/50.90206/12.12797 obviously it is much much slower than this one with 120kmh (https://www.openstreetmap.org/way/256772750#map=19/50.90260/12.12642)
![image](https://user-images.githubusercontent.com/1042025/89208973-b0cd4800-d5bd-11ea-9c64-1965d81c68bd.png)
